File: ClimaDaemon/Core/Clima.Core.Scheduler/Network/Services/LivestockService.cs
using Clima.Basics.Services.Communication;
using Clima.Core.Network.Messages;
using Clima.Core.Scheduler.Network.Messages;
using Clima.Core.Scheduler;

namespace Clima.Core.Scheduler.Network.Services
{
    public class LivestockService:INetworkService
    {
        private readonly IClimaScheduler _scheduler;

        public LivestockService(IClimaScheduler scheduler)
        {
            _scheduler = scheduler;
        }

        [ServiceMethod]
        public LivestockStateResponse PlantHeads(LivestockOperationRequest request)
        {
            _scheduler.LivestockPlanting(request.HeadsCount, request.OperationDate);
            return new LivestockStateResponse()
            {
                State = _scheduler.GetLivestockState()
            };
        }
        [ServiceMethod]
        public LivestockStateResponse KillHeads(LivestockOperationRequest request)
        {
            _scheduler.LivestockKill(request.HeadsCount, request.OperationDate);
            return new LivestockStateResponse()
            {
                State = _scheduler.GetLivestockState()
            };
        }
        [ServiceMethod]
        public LivestockStateResponse DeathHeads(LivestockOperationRequest request)
        {
            _scheduler.LivestockDeath(request.HeadsCount, request.OperationDate);
            return new LivestockStateResponse()
            {
                State = _scheduler.GetLivestockState()
            };
        }
        [ServiceMethod]
        public LivestockStateResponse RefractHeads(LivestockOperationRequest request)
        {
            _scheduler.LivestockRefraction(request.HeadsCount, request.OperationDate);
            return new LivestockStateResponse()
            {
                State = _scheduler.GetLivestockState()
            };
        }
        [ServiceMethod]
        public LivestockStateResponse GetLivestockState(DefaultRequest request)
        {
            return new LivestockStateResponse()
            {
                State = _scheduler.GetLivestockState()
            };
        }
        
    }
}

File: QtClient/QtClimaClient/Network/GenericServices/LivestockService.cpp
#include "LivestockService.h"

#include <Network/GenericServices/Messages/DefaultRequest.h>
#include <Network/GenericServices/Messages/LivestockOpListResponse.h>
#include <Network/GenericServices/Messages/LivestockOperationRequest.h>
#include <Network/GenericServices/Messages/LivestockStateResponse.h>

LivestockService::LivestockService(QObject *parent) : INetworkService(parent)
{

}

void LivestockService::GetLivestockState()
{
    NetworkRequest *request = new NetworkRequest();

    request->jsonrpc = "0.1a";
    request->service = "LivestockService";
    request->method = "GetLivestockState";

    emit SendRequest(request);
}

void LivestockService::Plant(const int &heads, const QDateTime &date)
{
    NetworkRequest *request = new NetworkRequest();

    request->jsonrpc = "0.1a";
    request->service = "LivestockService";
    request->method = "PlantHeads";

    LivestockOperationRequest op;
    op.HeadsCount = heads;
    op.OperationDate = date;

    request->params = op.toJsonString();

    emit SendRequest(request);
}

void LivestockService::Kill(const int &heads, const QDateTime &date)
{
    NetworkRequest *request = new NetworkRequest();

    request->jsonrpc = "0.1a";
    request->service = "LivestockService";
    request->method = "KillHeads";

    LivestockOperationRequest op;
    op.HeadsCount = heads;
    op.OperationDate = date;

    request->params = op.toJsonString();

    emit SendRequest(request);
}

void LivestockService::Refraction(const int &heads, const QDateTime &date)
{
    NetworkRequest *request = new NetworkRequest();

    request->jsonrpc = "0.1a";
    request->service = "LivestockService";
    request->method = "RefractHeads";
    LivestockOperationRequest op;
    op.HeadsCount = heads;
    op.OperationDate = date;

    request->params = op.toJsonString();

    emit SendRequest(request);
}

void LivestockService::Death(const int &heads, const QDateTime &date)
{
    NetworkRequest *request = new NetworkRequest();

    request->jsonrpc = "0.1a";
    request->service = "LivestockService";
    request->method = "DeathHeads";
    LivestockOperationRequest op;
    op.HeadsCount = heads;
    op.OperationDate = date;

    request->params = op.toJsonString();

    emit SendRequest(request);
}

void LivestockService::GetOpList()
{
    NetworkRequest *request = new NetworkRequest();

    request->jsonrpc = "0.1a";
    request->service = "LivestockService";
    request->method = "GetOperations";
    request->params = DefaultRequest().toJsonString();

    emit SendRequest(request);
}

void LivestockService::ProcessReply(const NetworkResponse &reply)
{
    if(reply.service == "LivestockService")
    {
        LivestockStateResponse resp;
        resp.fromJson(reply.result.toUtf8());

        if(reply.method == "GetLivestockState")
        {
            emit LivestockStateReceived(resp.State);
        }
        else if(reply.method == "KillHeads")
        {
            emit KillComlete();
        }
        else if(reply.method == "RefractHeads")
        {
            emit RefractionComplete();
        }
        else if(reply.method == "DeathHeads")
        {
            emit DeathComplete();
        }
        else if(reply.method == "PlantHeads")
        {
            emit PlantedComplete();
        }
        else if(reply.method == "GetOperations")
        {
            LivestockOpListResponse resp;
            resp.fromJson(reply.result.toUtf8());
            emit OpListReceived(resp.Operations);
        }

        emit LivestockUpdated(resp.State);
    }
}
